### PR TITLE
DropdownMenu: use KeyboardEvent.code, refactor tests to model RTL and user-event

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Refactor the `DropdownMenu ` component to rely on `code` instead of `keyCode` for keyboard events.

Since some tests were failing because they were relying on `fireEvent`, I went ahead and refactored them to using `userEvent` and with "modern" RTL style (i.e. avoiding implementation details).

## Why?

[`keyCode`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) is deprecated, and replaced by [`code`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code)

## How?

Easy swap of values

## Testing Instructions

In Storybook, visit the `DropdownMenu` story and make sure that the dropdown opens when focusing its toggle button and pressing the arrow down key.

### CODE REVIEW

**Comment by ciampo** [comment]
Pinging @geriux and @dcalhoun as this PR makes changes to a native component

File: packages/components/CHANGELOG.md
Comment (tyxla) [inline packages/components/CHANGELOG.md]:
```suggestion
-   `DropdownMenu`: use `KeyboardEvent.code`, refactor tests to model RTL and `user-event` ([#43439](https://github.com/WordPress/gutenberg/pull/43439/)).
```

File: packages/components/src/dropdown-menu/test/index.js
Comment (tyxla) [inline packages/components/src/dropdown-menu/test/index.js]: Noting that we no longer care about whether the `menuitem` is within the above `menu` - is that a problem? In the previous version of the tests, we were looking for menuitems only within the menu. This is achievable with `{within} from '@testing-library/dom'` for the protocol.

**Review by tyxla — APPROVED** [review]
Looks good 👍 Needs a rebase, otherwise ready to ship IMO 🚀

File: packages/components/src/dropdown-menu/test/index.js
Comment (ciampo) [reply packages/components/src/dropdown-menu/test/index.js]: Great call, I pushed https://github.com/WordPress/gutenberg/pull/43439/commits/a50dc67a6d60bf5349abe9d4298d030c2077b6df — let me know if this is what you had in mind!

Comment (tyxla) [reply packages/components/src/dropdown-menu/test/index.js]: Exactly what I meant - thanks 👍

**Comment by ciampo** [comment]
> this PR makes changes to a native component

Going to wait for CI checks to go ✅  and will merge, unless @geriux and/or @dcalhoun would like to review the code too.

**Comment by dcalhoun** [comment]
> Going to wait for CI checks to go ✅ and will merge, unless @geriux and/or @dcalhoun would like to review the code too.

👋🏻 @ciampo. I would be happy to review/test — it is on my todo list — but I made need a couple of days before I can dedicate time to this. I went ahead and triggered additional mobile CI tests.

**Comment by ciampo** [comment]
> > Going to wait for CI checks to go ✅ and will merge, unless @geriux and/or @dcalhoun would like to review the code too.
> 
> 👋🏻 @ciampo. I would be happy to review/test — it is on my todo list — but I made need a couple of days before I can dedicate time to this. I went ahead and triggered additional mobile CI tests.

That's great to hear! I'll wait for your review, then (this is not a high-priority PR anyway, so no need to rush!)

File: packages/components/src/dropdown-menu/index.native.js
Comment (dcalhoun) [inline packages/components/src/dropdown-menu/index.native.js]: From researching this a bit, I believe this is unused code. It is likely safe to remove `onKeyDown` and related code entirely from this file. 

For web, the `onKeyDown` is passed along via [prop spread](https://github.com/WordPress/gutenberg/blob/bf22040dc217a6846a0379e07e350ea843660c08/packages/components/src/button/index.js#L154). For native, we [do not prop spread](https://github.com/WordPress/gutenberg/blob/bf22040dc217a6846a0379e07e350ea843660c08/packages/components/src/button/index.native.js#L177-L186) `onKeyDown`, also it is not a valid prop for React Native's [`TouchableWithoutFeedback`](https://reactnative.dev/docs/0.66/touchablewithoutfeedback). 

* `onBlur`
* `onFocus`
* `onLayout`
* `onLongPress`
* `onPress`
* `onPressIn`
* `onPressOut`

All that said, it is safe to ship this as-is. The unused code is not necessarily related to this work.

**Review by dcalhoun — APPROVED** [review]
🚀 LGTM. I left one comment for context, but it is not a blocker. 

I tested the _Heading level_, _Alignment_ (horizontal and vertical), and _Justification_ dropdown menus on an iPhone SE and Samsung Galaxy S20. I believe those are the only locations leveraging `DropdownMenu` currently. 

Thank you for pinging the native team. 🙇🏻

File: packages/components/src/dropdown-menu/index.native.js
Comment (ciampo) [reply packages/components/src/dropdown-menu/index.native.js]: Since I had to rebase anyway to solve conflicts, I went ahead and removed the `onKeyDown` prop in https://github.com/WordPress/gutenberg/pull/43439/commits/f461f65f29b327903a3d51c80b67ad6f4910ea20

Let me know if the change look good to you!

Comment (ciampo) [reply packages/components/src/dropdown-menu/index.native.js]: Actually I noticed that some tests are failing in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5109.

I've reverted the removal of `onKeyDown`, let's see if it makes any difference

Comment (fluiddot) [reply packages/components/src/dropdown-menu/index.native.js]: I think some of the errors in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5109 should be solved with changes from https://github.com/WordPress/gutenberg/pull/43521, at least the ones affecting `Gutenberg Editor Slash Inserter tests should insert an embed image block with "/img" + enter` test case ([reference](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/18368/workflows/52857fc8-29bf-4d81-8efa-975cbac61a8c/jobs/100679/tests#failed-test-0)).

Comment (ciampo) [reply packages/components/src/dropdown-menu/index.native.js]: Thank you for the context — I'll wait for it to get merged and take a look at the tests again

Comment (dcalhoun) [reply packages/components/src/dropdown-menu/index.native.js]: https://github.com/WordPress/gutenberg/pull/43439/commits/f461f65f29b327903a3d51c80b67ad6f4910ea20 looks sound to me. I agree with @fluiddot that the CI failures are unrelated to this PR. Hopefully we validate that once https://github.com/WordPress/gutenberg/pull/43521 is merged and this branch is rebased.

Comment (ciampo) [reply packages/components/src/dropdown-menu/index.native.js]: As https://github.com/WordPress/gutenberg/pull/43521 got merged, I rebased and restored the removal of the `onKeyDown` callback.

I'll wait for your final ✅  before merging (and will keep an eye for unit test failures)

**Review by dcalhoun — APPROVED** [review]

**Pull request merged**
